daemon/hooks/hooks.go

```diff
@@ -0,0 +1,125 @@
+package hooks
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+	"os/exec"
+	"path/filepath"
+	"sync"
+	"bufio"
+	"bytes"
+
+	"github.com/zrepl/zrepl/logger"
+)
+
+type contextKey int
+
+const (
+	contextKeyLog contextKey = 0
+)
+
+type Logger = logger.Logger
+
+func WithLogger(ctx context.Context, log Logger) context.Context {
+	return context.WithValue(ctx, contextKeyLog, log)
+}
+
+func getLogger(ctx context.Context) Logger {
+	if log, ok := ctx.Value(contextKeyLog).(Logger); ok {
+		return log
+	}
+	return logger.NewNullLogger()
+}
+
+type logWriter struct {
+	mu *sync.Mutex
+	buf bytes.Buffer
+	scanner *bufio.Scanner
+	logFunc func(string)
+	level logger.Level
+}
+
+func NewLogWriter(mu *sync.Mutex, logFunc func(string)) *logWriter {
+	w := new(logWriter)
+	w.mu = mu
+	w.scanner = bufio.NewScanner(&w.buf)
+	w.logFunc = logFunc
```

Comment:
I would prefer the use of structured logging for this, i.e. a proper `github.com/zrepl/zrepl/logger.Logger`.
Called with `l.WithField("std{out|err}", line).Log(level, "hook output")`
Feel free to extend the `logger.Logger` interface with such a `Log(logger.Level, string)` method.